File: DesignPatternDemo/Pattern/Visitor/Visitor.m
//
//  Visitor.m
//  DesignPatternDemo
//
//  Created by ruantong on 2019/5/11.
//  Copyright © 2019 wupeng. All rights reserved.
//

#import "Visitor.h"
#import "Manager.h"
#import "CommonEmployee.h"

@implementation Visitor

- (void)visitCommonEmployee:(CommonEmployee *)commonEmployee{
    NSLog(@"%@",[self getCommonEmployeeInfo:commonEmployee]);
}

- (void)visitManager:(Manager *)manager{
    NSLog(@"%@",[self getManagerInfo:manager]);
}

- (NSString *)getBasicInfo:(Employee *)employee{
    NSString * info = [NSString stringWithFormat: @"姓名:%@",employee.name];
    info = [NSString stringWithFormat:@"%@ 性别:%@",info,employee.sex == 1?@"女":@"男"];
    info = [NSString stringWithFormat:@"%@ 薪水:%ld",info,employee.salary];
    return info;
}

- (NSString *)getManagerInfo:(Manager *)manager{
    NSString * info = [self getBasicInfo:manager];
    info = [NSString stringWithFormat:@"%@ 业绩:%@",info,manager.getOtherInfo];
    return info;
}

- (NSString *)getCommonEmployeeInfo:(CommonEmployee *)commonEmployee{
    NSString * info = [self getBasicInfo:commonEmployee];
    info = [NSString stringWithFormat:@"%@ 工作:%@",info,commonEmployee.job];
    return info;
}

@end

File: DesignPatternDemo/Pattern/Visitor/Visitor.swift
//
//  Visitor.swift
//  DesignPatternDemo
//
//  Created by wupeng on 2020/3/4.
//  Copyright © 2020 wupeng. All rights reserved.
//

import Foundation

protocol IVisitor{
    func visitCommonEmployee(commonEmployee:CommonEmployee);
    func visitManager(manager:Manager);
}

class Visitor: IVisitor {
    
    func visitManager(manager: Manager) {
        let info = getManagerInfo(manager: manager)
        print(info)
    }
    
    func visitCommonEmployee(commonEmployee: CommonEmployee) {
        let info = getCommonEmployeeInfo(commonEmployee: commonEmployee)
        print(info)
    }
    
    func getBasicInfo(employee:Employee) -> NSString {
        let info = "姓名:\(employee.name) 性别:\(employee.sex) 薪水:\(employee.salary)" as NSString
        return info
    }
    
    func getManagerInfo(manager:Manager) -> NSString {
        var info = getBasicInfo(employee: manager)
        info = "\(info) 业绩:\(manager.performance)" as NSString
        return info
    }
    
    func getCommonEmployeeInfo(commonEmployee:CommonEmployee) -> NSString {
        var info = getBasicInfo(employee: commonEmployee)
        info = "\(info) 工作:\(commonEmployee.job)" as NSString
        return info
    }
}



